# Trabajo Práctico 3: Predicción de tormenta en base de Datos Meteorológicos usando modelo más avanzados

### Integrantes

- Andrioli, Facundo
- Pérez, José

### Modelado y Evaluación:

Utilizaremos el mismo dataset que en el TP2, manteniendo el mismo conjunto de entrenamiento y de evaluación, así como 
las métricas de evaluación definidas anteriormente. Seleccionar uno de los siguientes clasificadores vistos en las 
clases 6 y 7:
- SVC
- Árbol de clasificación
- Bosques aleatorios
- CatBoost

1. Evalúe el modelo con las métricas definidas. 

In [2]:
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix, f1_score
import pickle

# Cargamos los datos de entrenamiento y prueba guardados en TP2
X_train = pd.read_csv('X_train.csv')
X_test = pd.read_csv('X_test.csv')
y_train = pd.read_csv('y_train.csv')
y_test = pd.read_csv('y_test.csv')

# Entrenamos el modelo de Bosques Aleatorios
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train.values.ravel())  # ravel para convertir DataFrame a array 1D# Hacemos predicciones en el conjunto de prueba
y_pred_rf = rf_model.predict(X_test)

# Evaluamos el modelo
accuracy_rf = accuracy_score(y_test, y_pred_rf)
f1score_rf = f1_score(y_test, y_pred_rf, average='macro')
conf_matrix_rf = confusion_matrix(y_test, y_pred_rf)

# Mostramos los resultadosprint("\n--Modelo Bosques Aleatorios--")
print(f"Accuracy: {accuracy_rf}")
print(f"F1-score: {f1score_rf}")
print("Matriz de confusión del modelo:")
print(conf_matrix_rf)


Accuracy: 0.8506941187505862
F1-score: 0.7398598909208629
Matriz de confusión del modelo:
[[32056  1082]
 [ 5285  4221]]


In [3]:
# Guardamos el modelo en formato pickle para futuras evaluaciones
with open('random_forest_model.pkl', 'wb') as f:
    pickle.dump(rf_model, f)

2. Discuta las fortalezas y debilidades del modelo elegido en el contexto específico de este problema.
Se tiene la libertad de realizar búsqueda de hiperparámetros o utilizar los valores por defecto. En caso de usar SVC, se recomienda encarecidamente realizar una búsqueda de hiperparámetros.

El modelo de Bosques Aleatorios ofrece una buena precisión (85.07%) y maneja bien la complejidad de los datos, destacándose en problemas con muchas variables. Su capacidad para reducir el sobreajuste al combinar múltiples árboles lo convierte en una opción robusta. Sin embargo, su principal debilidad es la predicción de la clase minoritaria, como se ve en los falsos negativos (5,285). Aunque es menos interpretable que modelos como la regresión logística, su rendimiento justifica su uso en escenarios complejos como la predicción de tormentas. Además, puede ser computacionalmente intensivo, lo que podría ser una limitación en aplicaciones con recursos limitados.
Aparte se selecciono por su equilibrio entre rendimiento, robustez y facilidad de uso, lo que lo hace adecuado para este problema sin requerir una optimización intensiva de hiperparametro.

### Conclusiones:

1. Construya una tabla comparativa con los resultados de los diferentes modelos y el modelo base. ¿Cuál fue el mejor modelo y por qué?

2. Resuma los hallazgos más relevantes obtenidos durante todo el proceso desde el TP1 al TP3.

3. Guarde en formato pickle el mejor modelo, seleccionado según su criterio.